ccmlib/common.py

```diff
@@ -76,6 +77,36 @@ class UnavailableSocketError(CCMError):
     pass
 
 
+class LogPatternToVersion(object):
+
+    def __init__(self, versions_to_patterns, default_pattern):
+        self.versions_to_patterns = versions_to_patterns
+        if "0" not in self.versions_to_patterns:
```

Comment:
This is clever... a little too clever :P The problem here is that the fallback becomes emergent behavior, rather than explicit behavior. If we were writing some kind of high-performance code, that'd be worth the loss in readability, but I don't think it is here. I'd add a `default` attribute to the object itself and use that throughout. the `__repr__` should account for that.

And sorry to waffle on this, but I think that `default` should be an optional kwarg. I don't really like requiring callers to specify both a dict and a fallback, even in cases when they know they won't need the fallback. In cases where `__call__` would fall back and `default` wasn't specified, `__call__` should blow up.

At this point, I think that `__iter__` and `__len__` aren't particularly meaningful for this object, since it's not really a simple map, and I'm not sure what I'd use `patterns` and `versions` for. As to whether `versions` should return a list of `LooseVersions` or not -- I think the thing we should do is wait to see if we need it and then write the method up to that need.